app/backfill-eventId.js

```diff
@@ -0,0 +1,125 @@
+      return `${instrument} ${flag.toUpperCase()}${id.toLowerCase()}`
+    },
+  ],
+  [/ANTARES[.\s_-]*(\d{6}[a-z])/i, ([, id]) => `ANTARES ${id}`.toUpperCase()],
+  [
+    /Baksan\sNeutrino\sObservatory\sAlert[.\s_-]*(\d{6}.\d{2})/i,
+    ([, id]) => `Baksan Neutrino Observatory Alert ${id}`,
+  ],
+  [/EP[.\s_-]*(\d{6}[a-z])/i, ([, id]) => `EP${id}`],
+  [/FRB[.\s_-]*(\d{8}[a-z])/i, ([, id]) => `FRB ${id}`.toUpperCase()],
+]
+
+export function parseEventFromSubject(value) {
+  for (const [regexp, normalize] of subjectMatchers) {
+    const startsWithMatch = RegExp(`^${regexp.source}`).exec(value)
+    if (startsWithMatch) return normalize(startsWithMatch)
+  }
+  for (const [regexp, normalize] of subjectMatchers) {
+    const match = regexp.exec(value)
+    if (match) return normalize(match)
+  }
+}
+// END CUT AND PASTE OF APP CODE FROM /routes/circulars/circulars.lib.ts
+
+async function getTableNameFromSSM(dynamoTableName) {
+  const ssmClient = new SSMClient({ region: 'us-east-1' })
+
+  try {
+    const command = new GetParameterCommand({ Name: dynamoTableName })
+    const response = await ssmClient.send(command)
+
+    if (!response.Parameter?.Value) {
+      throw new Error('dynamoTableName not found in SSM')
+    }
+
+    return response.Parameter.Value
+  } catch (error) {
+    console.error('Error fetching table name from SSM:', error)
+    throw error
+  }
+}
+
+export async function backfillEventIds() {
+  const startTime = new Date()
+  console.log('Starting EVENT ID backfill...', startTime)
+  const writeLimit = 1000
+  const dynamoTableName = '/RemixGcnProduction/tables/circulars'
+  const TableName = await getTableNameFromSSM(dynamoTableName)
+  const client = new DynamoDBClient({ region: 'us-east-1' })
+  const pages = paginateScan({ client }, { TableName })
+  let totalWriteCount = 0
+  let limitHit = false
+
+  for await (const page of pages) {
+    if (limitHit) break
+    const chunked = chunk(page.Items || [], 25)
+
+    for (const chunk of chunked) {
+      if (limitHit) break
+      let writes = []
+
+      for (const record of chunk) {
+        if (limitHit) break
+
+        const circular = unmarshall(record)
+        const parsedEventId = parseEventFromSubject(circular.subject)
+
+        if (!circular.eventId && parsedEventId) {
+          circular.eventId = parsedEventId
+          writes.push(circular)
+        }
+      }
+
+      if (writes.length + totalWriteCount > writeLimit) {
+        const overage = writes.length + totalWriteCount - writeLimit
+        const writesToLimitTo = writes.length - overage
+        writes = writes.slice(0, writesToLimitTo)
+        limitHit = true
+      }
+
+      if (writes.length > 0) {
+        const command = new BatchWriteCommand({
+          RequestItems: {
+            [TableName]: writes.map((circ) => ({
+              PutRequest: {
+                Item: {
+                  ...circ,
+                },
+              },
+            })),
+          },
+        })
```

Comment:
This is quite complicated. I would be more comfortable with the correctness of a functional programming implementation using [TC39 Async Iterator Helpers](https://github.com/tc39/proposal-async-iterator-helpers) (see polyfills [available from core-js](https://github.com/zloirock/core-js?tab=readme-ov-file#asynciterator-helpers)).

Reply:
There are a lot of loops and conditionals here, which makes the code too complicated to review quickly.

As for whether we have existing code like this, I don't know of any other examples where we have a paginated input, which is filtered, and then which must be paginated again for insertion. That is the source of the complexity.

Reply:
So, the core-js polyfills would work just fine, but they won't be necessary if we make the following changes:

- Overwrite all event IDs, regardless of whether or not they are already populated.
- Process all events at once. No need to do it in small batches spread over days.

I discussed both of these changes with @jracusin, and she agrees.